velox/common/memory/SharedArbitrator.cpp

```diff
@@ -169,67 +146,157 @@ const SharedArbitrator::Candidate& findCandidateWithLargestCapacity(
 } // namespace
 
 SharedArbitrator::SharedArbitrator(const MemoryArbitrator::Config& config)
-    : MemoryArbitrator(config), freeCapacity_(capacity_) {
-  RECORD_METRIC_VALUE(kMetricArbitratorFreeCapacityBytes, freeCapacity_);
+    : MemoryArbitrator(config),
+      freeReservedCapacity_(reservedCapacity_),
+      freeNonReservedCapacity_(capacity_ - freeReservedCapacity_) {
   VELOX_CHECK_EQ(kind_, config.kind);
+  updateFreeCapacityMetrics();
 }
 
 std::string SharedArbitrator::Candidate::toString() const {
   return fmt::format(
-      "CANDIDATE[{} RECLAIMABLE[{}] RECLAIMABLE_BYTES[{}] FREE_BYTES[{}]]",
+      "CANDIDATE[{}] RECLAIMABLE_BYTES[{}] FREE_BYTES[{}]]",
       pool->root()->name(),
-      reclaimable,
       succinctBytes(reclaimableBytes),
       succinctBytes(freeBytes));
 }
 
 SharedArbitrator::~SharedArbitrator() {
-  if (freeCapacity_ != capacity_) {
+  if (freeReservedCapacity_ != reservedCapacity_) {
+    const std::string errMsg = fmt::format(
+        "There is unexpected free reserved capacity not given back to arbitrator "
+        "on destruction: freeReservedCapacity_{} != reservedCapacity_{}\\n{}",
+        freeReservedCapacity_,
+        reservedCapacity_,
+        toString());
+    if (checkUsageLeak_) {
+      VELOX_FAIL(errMsg);
+    } else {
+      VELOX_MEM_LOG(ERROR) << errMsg;
+    }
+  }
+  if (freeNonReservedCapacity_ + freeReservedCapacity_ != capacity_) {
     const std::string errMsg = fmt::format(
-        "\"There is unexpected free capacity not given back to arbitrator "
-        "on destruction: freeCapacity_ != capacity_ ({} vs {})\\n{}\"",
-        freeCapacity_,
+        "There is unexpected free capacity not given back to arbitrator "
+        "on destruction: freeNonReservedCapacity_[{}] + freeReservedCapacity_[{}] != capacity_[{}])\\n{}",
+        freeNonReservedCapacity_,
+        freeReservedCapacity_,
         capacity_,
         toString());
     if (checkUsageLeak_) {
       VELOX_FAIL(errMsg);
     } else {
-      LOG(ERROR) << errMsg;
+      VELOX_MEM_LOG(ERROR) << errMsg;
     }
   }
 }
 
+std::vector<SharedArbitrator::Candidate> SharedArbitrator::getCandidateStats(
+    const std::vector<std::shared_ptr<MemoryPool>>& pools,
+    bool freeCapacityOnly) {
+  std::vector<SharedArbitrator::Candidate> candidates;
+  candidates.reserve(pools.size());
+  for (const auto& pool : pools) {
+    candidates.push_back(
+        {freeCapacityOnly ? 0 : reclaimableUsedCapacity(*pool),
+         reclaimableFreeCapacity(*pool),
+         pool->currentBytes(),
+         pool.get()});
+  }
+  return candidates;
+}
+
+void SharedArbitrator::updateFreeCapacityMetrics() const {
+  RECORD_METRIC_VALUE(
+      kMetricArbitratorFreeCapacityBytes,
+      freeNonReservedCapacity_ + freeReservedCapacity_);
+  RECORD_METRIC_VALUE(
+      kMetricArbitratorFreeReservedCapacityBytes, freeReservedCapacity_);
```

Comment:
Should we export freeReservedCapacity and freeNonReservedCapacity instead of a combined total? That way using metrics view tools it will be more intuitive to manipulate around.